concordium-smart-contract-testing/src/impls.rs

```diff
@@ -0,0 +1,962 @@
+                        energy_used,
+                        transaction_fee,
+                    }),
+                    transaction_fee,
+                )
+            }
+            Ok(v1::InitResult::OutOfEnergy) => {
+                let transaction_fee = self.calculate_energy_cost(energy_reserved);
+                (
+                    Err(ContractInitError::ExecutionError {
+                        failure_kind: InitFailure::OutOfEnergy,
+                        energy_used: energy_reserved,
+                        transaction_fee,
+                    }),
+                    transaction_fee,
+                )
+            }
+            Err(e) => panic!("Internal error: init failed with interpreter error: {}", e),
+        };
+        // Charge the account.
+        // We have to get the account info again because of the borrow checker.
+        self.account_mut(sender)?.balance.total -= transaction_fee;
+        res
+    }
+
+    /// Update a contract by calling one of its entrypoints.
+    ///
+    /// If successful, all changes will be saved.
+    ///
+    /// **Parameters:**
+    ///  - `invoker`: the account paying for the transaction.
+    ///  - `sender`: the sender of the transaction, can also be a contract.
+    ///  - `contract_address`: the contract to update.
+    ///  - `entrypoint`: the entrypoint to call.
+    ///  - `parameter`: the contract parameter.
+    ///  - `amount`: the amount sent to the contract.
+    ///  - `energy_reserved`: the maximum energy that can be used in the update.
+    pub fn contract_update(
+        &mut self,
+        invoker: AccountAddress,
+        sender: Address,
+        contract_address: ContractAddress,
+        entrypoint: EntrypointName,
+        parameter: OwnedParameter,
+        amount: Amount,
+        energy_reserved: Energy,
+    ) -> Result<SuccessfulContractUpdate, ContractUpdateError> {
+        // Ensure the sender exists.
+        if !self.address_exists(sender) {
+            return Err(ContractUpdateError::SenderDoesNotExist(sender));
+        }
+
+        let invoker_amount_reserved_for_nrg = self.calculate_energy_cost(energy_reserved);
+        // Ensure account exists and can pay for the reserved energy and amount
+        let account_info = self.account_mut(invoker)?;
+        let invoker_signature_count = account_info.signature_count;
+        if account_info.balance.available() < invoker_amount_reserved_for_nrg + amount {
+            return Err(ContractUpdateError::InsufficientFunds);
+        }
+
+        // Ensure that the parameter has a valid size (+2 for the length of parameter).
+        if parameter.0.len() + 2 > constants::MAX_PARAMETER_SIZE {
+            return Err(ContractUpdateError::ParameterTooLarge);
+        }
+
+        // Charge account for the reserved energy up front. This is to ensure that
+        // contract queries for the invoker balance are correct.
+        account_info.balance.total -= invoker_amount_reserved_for_nrg;
```

Comment:
No, because we never return early between these two points in the code. However, that is not documented and could easily break. I will revise how this is handled.